batch-coupon-exchang/src/main/kotlin/com/example/estdelivery/job/step/ExchangeCouponSteps.kt

```diff
@@ -0,0 +1,91 @@
+package com.example.estdelivery.job.step
+
+import com.example.estdelivery.ChunkSizeProperty
+import com.example.estdelivery.domain.CouponExchangeNotificationTemplate
+import com.example.estdelivery.domain.CouponStateAmountType
+import com.example.estdelivery.domain.MemberCoupon
+import com.example.estdelivery.job.step.service.CouponService
+import com.example.estdelivery.job.step.service.ExchangedCouponIdStorage
+import org.springframework.batch.core.configuration.annotation.JobScope
+import org.springframework.batch.core.repository.JobRepository
+import org.springframework.batch.core.step.builder.StepBuilder
+import org.springframework.batch.item.ItemProcessor
+import org.springframework.batch.item.ItemWriter
+import org.springframework.batch.item.database.JpaCursorItemReader
+import org.springframework.batch.item.database.JpaItemWriter
+import org.springframework.batch.repeat.RepeatStatus.FINISHED
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.context.annotation.Bean
+import org.springframework.context.annotation.Configuration
+import org.springframework.transaction.PlatformTransactionManager
+
+private const val CREATE_COUPON_TO_BE_EXCHANGE = "CREATE_COUPON_TO_BE_EXCHANGE"
+private const val EXCHANGE_COUPON = "EXCHANGE_COUPON"
+private const val EXCHANGE_COUPON_ALIM_TALK = "EXCHANGE_COUPON_ALIM_TALK"
+
+@Configuration
+class ExchangeCouponSteps {
+    @Bean
+    fun exchangeCouponStep(
+        jobRepository: JobRepository,
+        transactionManager: PlatformTransactionManager,
+        couponService: CouponService,
+        chunkSizeProperty: ChunkSizeProperty,
+        memberCouponReader: JpaCursorItemReader<MemberCoupon>,
+        exchangeCouponProcessor: ItemProcessor<MemberCoupon, MemberCoupon>,
+        updateMemberCouponWriter: JpaItemWriter<MemberCoupon>,
+    ) = StepBuilder(EXCHANGE_COUPON, jobRepository)
+        .chunk<MemberCoupon, MemberCoupon>(chunkSizeProperty.commitSize, transactionManager)
```

Comment:
`TransactionManager`가 영향을 주는 건 처리하는 데이터 말고도 Job Repository에 저장되는 데이터도 포함이 되는건가?

Reply:
그럴 수도 있고 그렇지 않을 수도 있다. 주석에 따르면 청크 기반 작업을 처리하기 위해 필요한 트랜잭션 매니저를 구성하는 방법이다. 

```
	/**
	 * Build a step that processes items in chunks with the size provided. To extend the
	 * step to being fault tolerant, call the {@link SimpleStepBuilder#faultTolerant()}
	 * method on the builder. In most cases you will want to parameterize your call to
	 * this method, to preserve the type safety of your readers and writers, e.g.
	 *
	 * <pre>
	 * new StepBuilder(&quot;step1&quot;).&lt;Order, Ledger&gt; chunk(100, transactionManager).reader(new OrderReader()).writer(new LedgerWriter())
	 * // ... etc.
	 * </pre>
	 * @param chunkSize the chunk size (commit interval)
	 * @param transactionManager the transaction manager to use for the chunk-oriented
	 * tasklet
	 * @return a {@link SimpleStepBuilder}
	 * @param <I> the type of item to be processed as input
	 * @param <O> the type of item to be output
	 * @since 5.0
	 */
	public <I, O> SimpleStepBuilder<I, O> chunk(int chunkSize, PlatformTransactionManager transactionManager) {
		return new SimpleStepBuilder<I, O>(this).transactionManager(transactionManager).chunk(chunkSize);
	}

```

https://github.com/spring-projects/spring-batch/issues/3950 해당 이슈를 보면 내부 동작을 더 잘 확인할 수 있다.

배치 관련 이력은 트랙잭션에 포함되지 않고 저장된다. 그다음 트랜잭션을 열어 작업을 실행한다. 작업 도 중 컨텍스트를 조회한다면 트랜잭션 내부에서 조회하는 일이기 때문에 트랜잭션에 포함된다.

작업 관련 외부 쿼리에서 동시성을 제어하기 위해 버전을 활용하며 JobExecution 의 synchronizeStatus를 이용해 버전을 관리한다. 관련 키워드도 함께 정리한다.

- JobExecution : 작업 정보가 포함된다. Job 식별자 등등..
- ExecutionContext : 작업이 어디까지 실행됐는지 등 ItemStream에서 활용하기 위해 저장된다.
- ItemStream : 오류가 발생할 경우 상태를 주기적으로 저장하고 해당 상태에서 복원하기 위한 계약을 정의하는 마커 인터페이스다.

Reply:
⭐ 다시 확인해보니 트랜잭션을 걸고 있다. AbstractJobRepositoryFactoryBean 에서 getObject 메서드를 호출하는 시점에 트랜잭션을 걸게 된다.